File: csharp/Modelos.cs
namespace AprendizajeIndividual;

/// <summary>
/// STRUCT: tipo de VALOR. Se copia al asignarlo o al pasarlo como parametro.
/// Sus campos son publicos y mutables, de modo que funciona como un simple
/// contenedor de datos relacionados, sin comportamiento propio.
/// </summary>
public struct EstudianteStruct
{
    public string Nombre;
    public int Edad;
    public double Promedio;

    public EstudianteStruct(string nombre, int edad, double promedio)
    {
        Nombre = nombre;
        Edad = edad;
        Promedio = promedio;
    }

    public override string ToString() =>
        $"{Nombre,-10} edad: {Edad,2}   promedio: {Promedio,5:F2}";
}

/// <summary>
/// RECORD: tipo de REFERENCIA inmutable con igualdad por valor.
/// Los campos se declaran en el encabezado (parametros posicionales) y no
/// pueden reasignarse: para "modificar" uno se crea una copia con la
/// expresion 'with'.
/// </summary>
public record EstudianteRecord(string Nombre, int Edad, double Promedio)
{
    public override string ToString() =>
        $"{Nombre,-10} edad: {Edad,2}   promedio: {Promedio,5:F2}";
}

/// <summary>
/// CLASE (objeto): tipo de REFERENCIA con estado encapsulado y comportamiento.
/// Los campos son privados y se exponen mediante metodos, lo que permite
/// validar los datos antes de aceptarlos.
/// </summary>
public class Estudiante
{
    private readonly string nombre;
    private readonly int edad;
    private double promedio;

    public Estudiante(string nombre, int edad, double promedio)
    {
        this.nombre = nombre;
        this.edad = edad;
        this.promedio = promedio;
    }

    public string Nombre => nombre;

    public double GetPromedio() => promedio;

    /// <summary>
    /// Cambia el promedio validando el rango. Esta validacion es justamente lo
    /// que un struct de campos publicos no puede garantizar.
    /// </summary>
    public void SetPromedio(double nuevoPromedio)
    {
        if (nuevoPromedio < 0.0 || nuevoPromedio > 5.0)
        {
            Console.WriteLine($"  [RECHAZADO] {nuevoPromedio:F2} esta fuera del rango 0.0 - 5.0");
            return;
        }
        promedio = nuevoPromedio;
    }

    /// <summary>Metodo de comportamiento exigido por el enunciado.</summary>
    public void MostrarInfo()
    {
        Console.WriteLine($"  {nombre,-10} edad: {edad,2}   promedio: {promedio,5:F2}");
    }
}

File: python/modelos.py
"""Los tres modelos del enunciado, en Python: registro mutable, registro
inmutable y objeto.

Python no tiene tipos de valor: toda variable guarda una referencia. Por eso no
existe un equivalente exacto del struct de C#. Lo que ofrece el lenguaje son
dos formas de declarar un registro:

    @dataclass              -> registro MUTABLE   (el papel del struct)
    NamedTuple / frozen     -> registro INMUTABLE (el papel del record)

La inmutabilidad de NamedTuple y de @dataclass(frozen=True) se impone en tiempo
de ejecucion: intentar asignar un campo lanza una excepcion, pero el objeto
sigue estando en el heap y sigue pasandose por referencia.
"""
from dataclasses import dataclass
from typing import NamedTuple


@dataclass
class EstudianteStruct:
    """Registro MUTABLE: el papel que cumple un struct de campos publicos.

    Es un contenedor de datos relacionados, sin comportamiento ni validacion.
    """
    nombre: str
    edad: int
    promedio: float

    def __str__(self) -> str:
        return f"{self.nombre:<10} edad: {self.edad:2}   promedio: {self.promedio:5.2f}"


class EstudianteRecord(NamedTuple):
    """Registro INMUTABLE con igualdad por valor: el papel del record.

    Para "modificar" un campo se crea una copia con el metodo _replace,
    equivalente a la expresion 'with' de C#.
    """
    nombre: str
    edad: int
    promedio: float

    def __str__(self) -> str:
        return f"{self.nombre:<10} edad: {self.edad:2}   promedio: {self.promedio:5.2f}"


class Estudiante:
    """OBJETO: estado encapsulado y comportamiento propio."""

    def __init__(self, nombre: str, edad: int, promedio: float) -> None:
        self._nombre = nombre
        self._edad = edad
        self._promedio = promedio

    @property
    def nombre(self) -> str:
        return self._nombre

    def get_promedio(self) -> float:
        return self._promedio

    def set_promedio(self, nuevo_promedio: float) -> None:
        """Cambia el promedio validando el rango.

        Esta validacion es justamente lo que un registro de campos publicos no
        puede garantizar.
        """
        if nuevo_promedio < 0.0 or nuevo_promedio > 5.0:
            print(f"  [RECHAZADO] {nuevo_promedio:.2f} esta fuera del rango 0.0 - 5.0")
            return
        self._promedio = nuevo_promedio

    def mostrar_info(self) -> None:
        """Metodo de comportamiento exigido por el enunciado."""
        print(f"  {self._nombre:<10} edad: {self._edad:2}   promedio: {self._promedio:5.2f}")
